packages/vite/src/node/ssr/runtime/types.ts

```diff
@@ -0,0 +1,185 @@
+   * @param id ID that was used to fetch the module
+   */
+  runViteModule(
+    context: ViteRuntimeModuleContext,
+    code: string,
+    id: string,
+  ): Promise<any>
+  /**
+   * Run externalized module.
+   * @param file File URL to the external module
+   */
+  runExternalModule(file: string): Promise<any>
+}
+
+export interface ModuleCache {
+  promise?: Promise<any>
+  exports?: any
+  evaluated?: boolean
+  map?: DecodedMap
+  meta?: FetchResult
+  /**
+   * Module ids that imports this module
+   */
+  importers?: Set<string>
+  imports?: Set<string>
+}
+
+export type FetchResult = ExternalFetchResult | ViteFetchResult
+
+export interface ExternalFetchResult {
+  /**
+   * The path to the externalized module starting with file://,
+   * by default this will be imported via a dynamic "import"
+   * instead of being transformed by vite and loaded with vite runtime
+   */
+  externalize: string
+  /**
+   * Type of the module. Will be used to determine if import statement is correct.
+   * For example, if Vite needs to throw an error if variable is not actually exported
+   */
+  type?: 'module' | 'commonjs' | 'builtin' | 'network'
+}
+
+export interface ViteFetchResult {
+  /**
+   * Code that will be evaluated by vite runtime
+   * by default this will be wrapped in an async function
+   */
+  code: string
+  /**
+   * File path of the module on disk.
+   * This will be resolved as import.meta.url/filename
+   */
+  file: string | null
+}
+
+export type ResolvedResult = (ExternalFetchResult | ViteFetchResult) & {
+  id: string
+}
+
+export type FetchFunction = (
+  id: string,
+  importer?: string,
+) => Promise<FetchResult>
+
+export interface ViteRuntimeOptions {
+  /**
+   * Root of the project
+   */
+  root: string
+  /**
+   * A method to get the information about the module.
+   * For SSR, Vite exposes `server.ssrFetchModule` function that you can use here.
+   * For other runtime use cases, Vite also exposes `fetchModule` from its main entry point.
+   */
+  fetchModule: FetchFunction
+  /**
+   * Custom environment variables available on `import.meta.env`. This doesn't modify the actual `process.env`.
+   */
+  environmentVariables?: Record<string, any>
+  /**
+   * Configure how source maps are resolved. Prefers `node` if `process.setSourceMapsEnabled` is available.
+   * Otherwise it will use `prepareStackTrace` by default.
+   */
+  sourcemapInterceptor?:
+    | false
+    | 'node'
+    | 'prepareStackTrace'
+    | InterceptorOptions
+  /**
+   * Disable HMR or configure HMR options.
+   */
+  hmr?:
+    | false
+    | {
+        /**
+         * Configure how HMR communicates between the client and the server.
+         */
+        connection: HMRRuntimeConnection
+        /**
+         * Configure HMR logger.
+         */
+        logger?: false | HMRLogger
+      }
+  /**
+   * Custom module cache. If not provided, creates a separate module cache for each ViteRuntime instance.
+   */
+  moduleCache?: ModuleCacheMap
```

Comment:
The `ModuleCacheMap` implementation is quite complex. Is there a possibility where someone could use this option and implement all its API? Or should we remove this option instead?

Reply:
It's possible to have several runtimes with a single module cache. `ModuleCacheMap` is exported and you can manually pass it down

Reply:
Ah ok so it's there so you can share the cache. Makes sense to me then 👍